File: apps/admin/src/tiebreaks.ts
/**
 * The tie-breaks a section Rochade pairs itself may rank by, in the engine's
 * own spelling. Mirrors `rochade.gacrux.tiebreaks` on the server: the codes
 * are what the section stores and what the standings table shows as column
 * headings, the labels are what the picker offers.
 */

export const TIEBREAKS: readonly { code: string; label: string }[] = [
  { code: "PTS", label: "Points" },
  { code: "BH/C1", label: "Buchholz Cut 1" },
  { code: "BH", label: "Buchholz" },
  { code: "BH/M1", label: "Median Buchholz" },
  { code: "SB", label: "Sonneborn-Berger" },
  { code: "SB/C1", label: "Sonneborn-Berger Cut 1" },
  { code: "DE", label: "Direct encounter" },
  { code: "WIN", label: "Wins" },
  { code: "WON", label: "Games won" },
  { code: "BPG", label: "Games with black" },
  { code: "BWG", label: "Wins with black" },
  { code: "PS", label: "Progressive score" },
  { code: "KS", label: "Koya" },
  { code: "AOB", label: "Average of opponents' Buchholz" },
  { code: "ARO", label: "Average rating of opponents" },
  { code: "ARO/C1", label: "Average rating of opponents Cut 1" },
  { code: "TPR", label: "Tournament performance rating" },
];

export const DEFAULT_TIEBREAKS: readonly string[] = ["PTS", "BH/C1", "BH", "SB"];

export const MAX_TIEBREAKS = 9;

export function tiebreakLabel(code: string): string {
  return TIEBREAKS.find((t) => t.code === code.toUpperCase())?.label ?? code;
}

File: src/rochade/gacrux/tiebreaks.py
"""The tie-breaks a section may rank by.

The engine knows some sixty codes, most of them for team events or for
checking other programs. This is the individual-Swiss subset an arbiter is
likely to want, spelled in the engine's own spec syntax so the section stores
exactly what it hands over, and labelled so the standings table has headings.

`PTS` always comes first: it is the score itself, and the engine's output
puts it in the first column, which is where `compute_standings` reads it.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Final

#: Spec -> label. Order is the order the picker offers them in.
KNOWN_TIEBREAKS: Final[dict[str, str]] = {
    "PTS": "Points",
    "BH/C1": "Buchholz Cut 1",
    "BH": "Buchholz",
    "BH/M1": "Median Buchholz",
    "SB": "Sonneborn-Berger",
    "SB/C1": "Sonneborn-Berger Cut 1",
    "DE": "Direct encounter",
    "WIN": "Wins",
    "WON": "Games won",
    "BPG": "Games with black",
    "BWG": "Wins with black",
    "PS": "Progressive score",
    "KS": "Koya",
    "AOB": "Average of opponents' Buchholz",
    "ARO": "Average rating of opponents",
    "ARO/C1": "Average rating of opponents Cut 1",
    "TPR": "Tournament performance rating",
}

#: FIDE's usual recommendation for an individual Swiss.
DEFAULT_TIEBREAKS: Final[tuple[str, ...]] = ("PTS", "BH/C1", "BH", "SB")

MAX_TIEBREAKS: Final = 9


def validate_tiebreaks(spec: Sequence[str]) -> list[str]:
    """Normalise a spec list, or raise `ValueError` saying what is wrong with it."""
    cleaned = [code.strip().upper() for code in spec]
    if not cleaned:
        raise ValueError("at least one tie-break is needed")
    if cleaned[0] != "PTS":
        raise ValueError("the first tie-break must be PTS, the points")
    if len(cleaned) > MAX_TIEBREAKS:
        raise ValueError(f"at most {MAX_TIEBREAKS} tie-breaks")
    unknown = [code for code in cleaned if code not in KNOWN_TIEBREAKS]
    if unknown:
        raise ValueError("unknown tie-break: " + ", ".join(sorted(set(unknown))))
    if len(set(cleaned)) != len(cleaned):
        raise ValueError("a tie-break is listed twice")
    return cleaned


def label_of(spec: str) -> str:
    return KNOWN_TIEBREAKS.get(spec.strip().upper(), spec)
